crates/fuel-core/src/graphql_api/storage/coins.rs

```diff
@@ -8,21 +8,253 @@ use fuel_core_storage::{
+        let bytes: [u8; MESSAGE_FOREIGN_KEY_LEN] =
+            value.foreign_key_bytes().try_into().map_err(|_| ())?;
+        Ok(fuel_types::Nonce::from(bytes))
+    }
+}
+
+impl CoinsToSpendIndexKey {
+    pub fn from_coin(coin: &Coin) -> Self {
+        let retryable_flag_bytes = NON_RETRYABLE_BYTE;
+        let address_bytes = coin.owner.as_ref();
+        let asset_id_bytes = coin.asset_id.as_ref();
+        let amount_bytes = coin.amount.to_be_bytes();
+        let utxo_id_bytes = utxo_id_to_bytes(&coin.utxo_id);
+
+        Self(
+            retryable_flag_bytes
+                .iter()
+                .chain(address_bytes)
+                .chain(asset_id_bytes)
+                .chain(amount_bytes.iter())
+                .chain(utxo_id_bytes.iter())
+                .copied()
+                .collect(),
+        )
+    }
+
+    pub fn from_message(message: &Message, base_asset_id: &AssetId) -> Self {
+        let retryable_flag_bytes = if message.is_retryable_message() {
+            RETRYABLE_BYTE
+        } else {
+            NON_RETRYABLE_BYTE
+        };
+        let address_bytes = message.recipient().as_ref();
+        let asset_id_bytes = base_asset_id.as_ref();
+        let amount_bytes = message.amount().to_be_bytes();
+        let nonce_bytes = message.nonce().as_slice();
+
+        Self(
+            retryable_flag_bytes
+                .iter()
+                .chain(address_bytes)
+                .chain(asset_id_bytes)
+                .chain(amount_bytes.iter())
+                .chain(nonce_bytes)
+                .copied()
+                .collect(),
+        )
+    }
+
+    fn from_slice(slice: &[u8]) -> Self {
+        Self(slice.into())
+    }
+
+    pub fn owner(&self) -> Address {
+        const ADDRESS_START: usize = RETRYABLE_FLAG_SIZE;
+        const ADDRESS_END: usize = ADDRESS_START + Address::LEN;
+        let address: [u8; Address::LEN] = self.0[ADDRESS_START..ADDRESS_END]
+            .try_into()
+            .expect("should have correct bytes");
+        Address::new(address)
+    }
+
+    pub fn asset_id(&self) -> AssetId {
+        const OFFSET: usize = RETRYABLE_FLAG_SIZE + Address::LEN;
+        const ASSET_ID_START: usize = OFFSET;
+        const ASSET_ID_END: usize = ASSET_ID_START + AssetId::LEN;
+        let asset_id: [u8; AssetId::LEN] = self.0[ASSET_ID_START..ASSET_ID_END]
+            .try_into()
+            .expect("should have correct bytes");
+        AssetId::new(asset_id)
+    }
+
+    pub fn retryable_flag(&self) -> u8 {
+        const OFFSET: usize = 0;
+        self.0[OFFSET]
+    }
+
+    pub fn amount(&self) -> u64 {
+        const OFFSET: usize = RETRYABLE_FLAG_SIZE + Address::LEN + AssetId::LEN;
+        const AMOUNT_START: usize = OFFSET;
+        const AMOUNT_END: usize = AMOUNT_START + AMOUNT_SIZE;
+        u64::from_be_bytes(
+            self.0[AMOUNT_START..AMOUNT_END]
+                .try_into()
+                .expect("should have correct bytes"),
```

Comment:
Not super happy about the expects, but I see the comment. However, I'm also not a fan of using the slice operator as it may cause unexpected panics. I'd prefer if we use `.get().expect("...")` explicitly instead?

```suggestion
            self.0.get(AMOUNT_START..AMOUNT_END)
                .expect("should span the whole range from AMOUNT_START to AMOUNT_END")
                .try_into()
                .expect("should have correct bytes"),
```

Reply:
Yeah, I wasn't super happy to so I converted the functions. Instead of panicking they will return `Option<T>`. Rest of the code is adjusted to handle the `None` case. Changes in https://github.com/FuelLabs/fuel-core/pull/2463/commits/792dc332f334c3e49b99f48b5c42633e185e4eb1

Reply:
Nice, looks great! Thanks for updating 🙏